pallets/pallet-asset-swap/src/xcm/trade.rs

```diff
@@ -109,56 +155,147 @@ mod swap_pair_remote_asset {
-			self.0 = self.0.saturating_add(weight);
-			self.1 = self.1.saturating_add(amount);
+
+			// Set link to XCM message ID only if this is the trader used.
+			log::trace!(target: LOG_TARGET, "Required {:?} - unused {:?}", required, unused);
+			self.consumed_xcm_hash = Some(context.message_id);
+			self.remaining_fungible_balance = self.remaining_fungible_balance.saturating_add(amount);
+			self.remaining_weight = self.remaining_weight.saturating_add(weight);
+
 			Ok(unused)
 		}
 
 		fn refund_weight(&mut self, weight: Weight, context: &XcmContext) -> Option<MultiAsset> {
 			log::trace!(target: LOG_TARGET, "UsingComponents::refund_weight weight: {:?}, context: {:?}", weight, context);
-			let swap_pair = SwapPair::<T>::get()?;
+
+			// Ensure we refund in the same trader we took fees from.
+			if Some(context.message_id) != self.consumed_xcm_hash {
+				return None;
+			};
+
+			let Some(ref swap_pair) = self.swap_pair else {
+				log::error!(target: LOG_TARGET, "Stored swap pair should not be None, but it is.");
+				return None;
+			};
+
 			let swap_pair_remote_asset_v3: AssetId = swap_pair.remote_asset_id.clone().try_into().map_err(|e| {
 				log::error!(target: LOG_TARGET, "Failed to convert stored asset ID {:?} into v3 AssetId with error {:?}", swap_pair.remote_asset_id, e);
 				Error::FailedToTransactAsset("Failed to convert swap pair asset ID into required version.")
 			}).ok()?;
-			let weight = weight.min(self.0);
+
+			let weight = weight.min(self.remaining_weight);
 			let amount = WeightToFee::weight_to_fee(&weight);
-			self.0 -= weight;
-			self.1 = self.1.saturating_sub(amount);
-			let amount: u128 = amount.saturated_into();
+
+			self.consumed_xcm_hash = None;
+			self.remaining_fungible_balance = self
+				.remaining_fungible_balance
+				.saturating_sub(self.remaining_fungible_balance);
+			self.remaining_weight = self.remaining_weight.saturating_sub(weight);
+
 			if amount > 0 {
-				log::trace!(target: LOG_TARGET, "refund amount {:?}", (swap_pair_remote_asset_v3, amount));
+				log::trace!(target: LOG_TARGET, "Refund amount {:?}", (swap_pair_remote_asset_v3, amount));
 				Some((swap_pair_remote_asset_v3, amount).into())
 			} else {
 				log::trace!(target: LOG_TARGET, "No refund");
 				None
 			}
 		}
 	}
+
+	// Move any unused asset from the swap pool account to the specified account,
+	// and update the remote balance with the difference since we know we control
+	// the full amount on the remote location.
+	impl<T, WeightToFee, FeeDestinationAccount> Drop
+		for UsingComponentsForSwapPairRemoteAsset<T, WeightToFee, FeeDestinationAccount>
+	where
+		T: Config,
+		FeeDestinationAccount: Get<T::AccountId>,
+	{
+		fn drop(&mut self) {
+			log::trace!(target: LOG_TARGET, "Drop with remaining {:?}", (self.consumed_xcm_hash, self.remaining_fungible_balance, self.remaining_weight, &self.swap_pair));
+			match (self.remaining_fungible_balance, &self.swap_pair) {
```

Comment:
Not a big fan of this match statement. I think a if is better suited

Reply:
https://github.com/KILTprotocol/kilt-node/pull/661/commits/ddff4eec5507f603d528e55c0ca2e2f0274cac94